x-pack/platform/plugins/shared/stack_connectors/server/connector_types/microsoft_defender_endpoint/o_auth_token_manager.ts

```diff
@@ -34,48 +39,127 @@ export class OAuthTokenManager {
     this.oAuthTokenUrl = url.toString();
   }
 
-  private async generateNewToken(connectorUsageCollector: ConnectorUsageCollector): Promise<void> {
-    // FYI: API Docs: https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#get-a-token
-    const { oAuthScope, clientId } = this.params.config;
-    const newToken = await this.params.apiRequest<MicrosoftDefenderEndpointApiTokenResponse>(
-      {
-        url: this.oAuthTokenUrl,
-        method: 'POST',
-        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
-        data: {
-          grant_type: 'client_credentials',
-          client_id: clientId,
-          scope: oAuthScope,
-          client_secret: this.params.secrets.clientSecret,
+  private isTokenExpired(token: ConnectorToken): boolean {
+    const now = new Date();
+    now.setSeconds(now.getSeconds() - 5); // Allows for a threshold of -5s before considering the token expired
+
+    const isExpired = token.expiresAt < now.toISOString();
+
+    if (isExpired) {
+      this.params.logger.debug(`Cached access token expired at [${token.expiresAt}]`);
+    }
+
+    return isExpired;
+  }
+
+  private async retrieveOrGenerateNewTokenIfNeeded(
+    connectorUsageCollector: ConnectorUsageCollector
+  ): Promise<void> {
+    if (this.generatingNewTokenPromise) {
+      return await this.generatingNewTokenPromise;
+    }
+
+    this.generatingNewTokenPromise = (async () => {
+      const {
+        connector: { id: connectorId },
+        logger,
+      } = this.params;
+      const connectorTokenClient = this.params.services.connectorTokenClient;
```

Comment:
is this `services.connectorTokenClient` a functionality exposed by response ops team, or you created that in a separate PR before?

Reply:
Yes, this is exposed by the ResponseOps framework. Its purpose is to only store tokens so they can be reused across instances of kibana.

Reply:
Thank you, I'll try to rebuild the token cache in CS to use this one 👍